RecoTracker/MkFitCore/src/IterationConfig.cc

```diff
@@ -99,14 +105,78 @@ namespace mkfit {
       /* vector<int> */ m_region_order,
       /* vector<mkfit::SteeringParams> */ m_steering_params,
       /* vector<mkfit::IterationLayerConfig> */ m_layer_configs
-      // /* function<void(const TrackerInfo&,const TrackVec&,const EventOfHits&,IterationSeedPartition&)> */   m_partition_seeds
   )
 
   ITCONF_DEFINE_TYPE_NON_INTRUSIVE(mkfit::IterationsInfo,
                                    /* vector<mkfit::IterationConfig> */ m_iterations)
 
   // End AUTO code.
 
+  // Begin IterationConfig function catalogs
+
+  namespace {
+    struct FuncCatalog {
+      std::map<std::string, IterationConfig::clean_seeds_func> seed_cleaners;
+      std::map<std::string, IterationConfig::partition_seeds_func> seed_partitioners;
+      std::map<std::string, IterationConfig::filter_candidates_func> candidate_filters;
+      std::map<std::string, IterationConfig::filter_duplicates_func> duplicate_cleaners;
+
+      std::mutex catalog_mutex;
+    };
+
+    FuncCatalog& get_catalog() {
+      static FuncCatalog func_catalog;
+      return func_catalog;
+    }
+  }
+
+  #define GET_FC auto& fc = get_catalog(); \
+                 const std::lock_guard<std::mutex> lock(fc.catalog_mutex)
+
+  void IterationConfig::register_seed_cleaner(const std::string& name, clean_seeds_func func) {
+    GET_FC;
+    fc.seed_cleaners.insert({name, func});
+  }
+  void IterationConfig::register_seed_partitioner(const std::string& name, partition_seeds_func func) {
+    GET_FC;
+    fc.seed_partitioners.insert({name, func});
+  }
+  void IterationConfig::register_candidate_filter(const std::string& name, filter_candidates_func func) {
+    GET_FC;
+    fc.candidate_filters.insert({name, func});
+  }
+  void IterationConfig::register_duplicate_cleaner(const std::string& name, filter_duplicates_func func) {
+    GET_FC;
+    fc.duplicate_cleaners.insert({name, func});
+  }
+
+  IterationConfig::clean_seeds_func IterationConfig::get_seed_cleaner(const std::string& name) {
+    GET_FC;
+    return fc.seed_cleaners[name];
```

Comment:
can the get be without a lock?
I guess using `.at(` may be implied in this case